### PR TITLE
[full-ci] Prepare batch actions in contextmenu

### LINKED ISSUE #5968: rightclick for multiple selected files
> As a user I want to have a contextmenu for multiple selected files to access the batchactions and details.

## Acceptance Criteria
- open the context menu for the current selction model
- dont reset the selectionmodel on rightclick
- available actions in the context menu
-- all available batch actions (context sensitive; download, copy, move, delete, accept share, decline share, restore)
-- details

-------

contexts are eg: "all files", "shared with me", "trashbin"

-------

### PR DESCRIPTION
**Pull request by pascalwengerter**
## Description
Perhaps could use updated unit/smoke tests

## Related Issue
- Fixes https://github.com/owncloud/web/issues/5968
- Fixes https://github.com/owncloud/web/issues/5977

## Open tasks
- [x] finish file actions mixins refactoring from single resources to multiple resources @kulmann

### CODE REVIEW

File: packages/web-app-files/src/components/BatchActions.vue [hunk 1]
```diff
@@ -1,118 +1,150 @@
 <template>
-  <div class="uk-flex uk-flex-middle">
+  <ul
```
Comment (kulmann) [inline packages/web-app-files/src/components/BatchActions.vue:2]: You could use the `oc-list` component here. Should do the same, but without UiKit.

File: packages/web-app-files/src/components/BatchActions.vue [hunk 2]
```diff
@@ -227,11 +259,24 @@ export default {
 
     isEmpty() {
       return this.activeFilesCurrentPage.length < 1
+    },
+
+    isDangerVariation() {
```
Comment (kulmann) [inline packages/web-app-files/src/components/BatchActions.vue:264]: The `is` in the computed prop name is confusing (same for the next two props), because you don't return a boolean but a string.

File: packages/web-app-files/src/components/BatchActions.vue [hunk 3]
```diff
@@ -1,118 +1,150 @@
...
           @click="declineShares()"
         >
           <oc-icon name="not_interested" />
           <translate>Decline</translate>
         </oc-button>
-      </div>
-    </oc-grid>
-  </div>
+      </li>
+      <li v-if="inContextActions && !isTrashbinRoute">
```
Comment (kulmann) [inline packages/web-app-files/src/components/BatchActions.vue:111]: You could use the `showDetails` action mixin here. You can copy the template from the ContextActions component and just use it here, it would avoid some code duplication.

File: packages/web-app-files/src/components/BatchActions.vue [hunk 4]
```diff
@@ -227,11 +259,24 @@ export default {
...
+    },
+
+    isPrimaryVariation() {
+      return this.inContextActions ? 'passive' : 'primary'
     }
   },
 
   methods: {
     ...mapActions('Files', ['removeFilesFromTrashbin', 'resetFileSelection']),
+    ...mapActions('Files/sidebar', { openSidebar: 'open' }),
```
Comment (kulmann) [inline packages/web-app-files/src/components/BatchActions.vue:279]: If you decide to use the mixin (as stated in another comment), this line would not be needed anymore - solved by the trigger method in the mixin.

File: packages/web-app-files/src/views/Favorites.vue [hunk 5]
```diff
@@ -25,7 +25,11 @@
           <quick-actions class="oc-visible@s" :item="props.resource" :actions="app.quickActions" />
         </template>
         <template #contextMenu="{ resource }">
-          <context-actions v-if="isHighlightedFile(resource)" :item="resource" />
+          <context-actions
+            v-if="!areMultipleSelected && isHighlightedFile(resource)"
+            :item="resource"
+          />
+          <batch-actions v-if="areMultipleSelected" :in-context-actions="true" />
```
Comment (kulmann) [inline packages/web-app-files/src/views/Favorites.vue:32]: Please also check if the clicked resource is part of the selected files. Otherwise you'd create the batch actions in the DOM tree for each and every row (same for the other views).

**Review by kulmann — CHANGES_REQUESTED** [review]
Hmmm. Good as a first solution. But I get the feeling that we should refactor the batch actions into `action` mixins as well and solve the batch-actions-context-menu inside the `ContextActions` component. That would also reduce business logic from the `BatchActions` component, which is always a good idea.

File: packages/web-app-files/src/components/BatchActions.vue (hunk 1 above)
Comment (pascalwengerter) [reply packages/web-app-files/src/components/BatchActions.vue:2]: `oc-list` also has zero margin (which will get interesting to sort out during the UiKit removal...)

File: packages/web-app-files/src/components/BatchActions.vue (hunk 2 above)
Comment (pascalwengerter) [reply packages/web-app-files/src/components/BatchActions.vue:264]: Good hint, thanks

File: packages/web-app-files/src/views/Favorites.vue (hunk 5 above)
Comment (pascalwengerter) [reply packages/web-app-files/src/views/Favorites.vue:32]: Uh yeah, great one!

File: packages/web-app-files/src/components/BatchActions.vue (hunk 3 above)
Comment (pascalwengerter) [reply packages/web-app-files/src/components/BatchActions.vue:111]: I initially did so, it turned out pretty ugly but let me take another look at it!

**Comment by pascalwengerter** [comment]
@kulmann can be re-based, I cherry-picked the ODS bump commit in #6009

File: packages/web-app-files/src/components/AppBar/AppBar.vue [hunk 6]
```diff
@@ -97,8 +98,8 @@
               </ul>
             </oc-drop>
           </template>
-          <size-info v-if="selectedFiles.length > 0" class="oc-mr-s uk-visible@l" />
-          <batch-actions />
+          <size-info v-if="hasBulkActions && selectedFiles.length > 0" class="oc-mr uk-visible@l" />
+          <batch-actions v-if="hasBulkActions" />
```
Comment (pascalwengerter) [inline packages/web-app-files/src/components/AppBar/AppBar.vue:102]: Unit tests compain and I'm not sure why batch.actions don't also get a `&& selectedFiles.length > 0` condition...I'll add it to make them pass (and test acceptance tests in CI)

Comment (kulmann) [reply packages/web-app-files/src/components/AppBar/AppBar.vue]: The batch actions are allowed without selected files now - there is a new action mixin in this PR (empty trash bin) which is only visible when no files are selected (on selection there is `delete` action mixin which handles permanent deletion of individually selected files). All action mixins now check if they are enabled with respect to the number of selected resources (some only available for 1 selected resource, some only available for >1 selected resources, some for >0 selected resources, one (empty trash) for 0 selected resources). With that the batch actions always need to be allowed (only can be suppressed with the route meta attribute `hasBulkActions`).

Long story short: we'll need to adapt the unit and integration tests accordingly.

File: packages/web-app-files/src/components/BatchActions.vue (hunk 1 above)
Comment (pascalwengerter) [reply packages/web-app-files/src/components/BatchActions.vue]: Pretty sure this has been resolved by you by now, marking as resolved

File: packages/web-app-files/src/components/BatchActions.vue (hunk 3 above)
Comment (pascalwengerter) [reply packages/web-app-files/src/components/BatchActions.vue]: Pretty sure this has been resolved by you by now, marking as resolved

File: packages/web-app-files/src/components/BatchActions.vue (hunk 4 above)
Comment (pascalwengerter) [reply packages/web-app-files/src/components/BatchActions.vue]: Pretty sure this has been resolved by you by now, marking as resolved

File: packages/web-app-files/src/views/Favorites.vue (hunk 5 above)
Comment (pascalwengerter) [reply packages/web-app-files/src/views/Favorites.vue]: Pretty sure this has been resolved by you (or was it me? 🤔 ) by now, marking as resolved

**Comment by ownclouders** [comment]
Results for <strong>oC10Files1</strong> https://drone.owncloud.com/owncloud/web/20290/15/1
The following scenarios passed on retry:
- webUIFilesActionMenu/versions.feature:15

**Comment by pascalwengerter** [comment]
Most recent CI run seems to have a lot of failures with modals not closing anymore, let's dig into that tomorrow!

**Comment by kulmann** [comment]
Rebased, since it has conflicts in the `fileActions.js` mixin

**Comment by ownclouders** [comment]
Results for <strong>oCISSharingPublic</strong> https://drone.owncloud.com/owncloud/web/20309/63/1
The following scenarios passed on retry:
- webUISharingPublicManagement/shareByPublicLink.feature:19

**Comment by ownclouders** [comment]
Results for <strong>oC10Files1</strong> https://drone.owncloud.com/owncloud/web/20322/15/1
The following scenarios passed on retry:
- webUIFilesActionMenu/versions.feature:15

File: tests/acceptance/features/webUIDeleteFilesFolders/deleteFilesFolders.feature [hunk 7]
```diff
@@ -248,7 +248,7 @@ Feature: deleting files and folders
     And as "Alice" file "lorem.txt" should not exist
     And no message should be displayed on the webUI
 
-
+  @skipOnOC10 @skipOnOCIS
```
Comment (individual-it) [inline tests/acceptance/features/webUIDeleteFilesFolders/deleteFilesFolders.feature:251]: complete scenario should be deleted, because its not a feature anymore

File: packages/web-app-files/src/mixins/actions/emptyTrashBin.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import { isTrashbinRoute } from '../../helpers/route'
+import { mapActions, mapGetters, mapState } from 'vuex'
+
+export default {
+  computed: {
+    ...mapState('Files', ['currentFolder']),
```
Comment (fschade) [inline packages/web-app-files/src/mixins/actions/emptyTrashBin.js:6]: currentFolder is not used

File: packages/web-app-files/src/mixins/actions/emptyTrashBin.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    ...mapActions('Files', ['clearTrashBin']),
+    $_emptyTrashBin_trigger() {
+      this.$client.fileTrash
+        .clearTrashBin()
+        .then(() => {
+          this.showMessage({
+            title: this.$gettext('All deleted files were removed')
+          })
+          this.clearTrashBin()
+        })
```
Comment (fschade) [inline packages/web-app-files/src/mixins/actions/emptyTrashBin.js:41]: can we use await catch syntax?

File: packages/web-app-files/src/mixins/actions/favorite.js [hunk 10]
```diff
@@ -38,10 +38,10 @@ export default {
   methods: {
     ...mapActions('Files', ['markFavorite']),
```
Comment (fschade) [inline packages/web-app-files/src/mixins/actions/favorite.js:39]: i would prefer to have await catch here

File: packages/web-app-files/src/mixins/actions/fetch.js [hunk 11]
```diff
@@ -1,24 +1,32 @@
-import MixinRoutes from '../routes'
 import { mapGetters } from 'vuex'
+import { isPublicFilesRoute } from '../../helpers/route'
 
 export default {
-  mixins: [MixinRoutes],
   computed: {
     ...mapGetters(['configuration']),
```
Comment (fschade) [inline packages/web-app-files/src/mixins/actions/fetch.js:6]: configuration isn't used

File: packages/web-app-files/src/mixins/actions/fetch.js [hunk 12]
```diff
@@ -29,14 +37,15 @@ export default {
...
+    $_fetch_trigger({ resources, mimeType, isPublicFile }) {
       if (isPublicFile) {
-        const url = file.downloadURL
+        const url = resources[0].downloadURL
         window.open(url, '_blank')
         return
       }
 
-      const url = this.$client.helpers._webdavUrl + file.path
+      // FIXME: use presigned URL instead
```
Comment (fschade) [inline packages/web-app-files/src/mixins/actions/fetch.js:47]: is this something for later or should happen in this pr?

File: packages/web-app-files/src/mixins/actions/fetch.js [hunk 13]
```diff
@@ -47,7 +56,7 @@ export default {
         headers
       })
         .then((r) => r.blob())
```
Comment (fschade) [inline packages/web-app-files/src/mixins/actions/fetch.js:58]: can we us async await here?

File: packages/web-app-files/tests/unit/components/AppBar/SelectedResources/BatchActions.spec.js [hunk 14]
```diff
@@ -15,369 +12,12 @@ const localVue = createLocalVue()
...
-        expect(spyRestoreFiles).toHaveBeenCalledTimes(1)
-      })
-      it('should call "$_deleteResources_displayDialog" when delete button is clicked', async () => {
-        await deleteButton.trigger('click')
-
-        expect(spyDeleteResourcesDisplayDialog).toHaveBeenCalledTimes(1)
-      })
-    })
-  })
+  it.todo('renders an empty list if there are no batch actions available')
```
Comment (fschade) [inline packages/web-app-files/tests/unit/components/AppBar/SelectedResources/BatchActions.spec.js:21]: is this something for a separate PR?

**Review by fschade — CHANGES_REQUESTED** [review]
really great changes, just found some monor things

File: packages/web-app-files/tests/unit/components/AppBar/SelectedResources/BatchActions.spec.js (hunk 14 above)
Comment (pascalwengerter) [reply packages/web-app-files/tests/unit/components/AppBar/SelectedResources/BatchActions.spec.js:21]: Yes

File: packages/web-app-files/src/mixins/actions/acceptShare.js [hunk 15]
```diff
@@ -30,22 +37,47 @@ export default {
...
-      } catch (error) {
-        console.error(error)
-        this.showMessage({
-          title: this.$gettext('Error while accepting the selected share.'),
-          status: 'danger'
-        })
+      })
+      await Promise.all(triggerPromises)
+
+      if (errors.length === 0) {
```
[the remark this answers is not in the document]
Comment (pascalwengerter) [reply packages/web-app-files/src/mixins/actions/acceptShare.js:68]: IMHO same-same in this case, no?

File: packages/web-app-files/src/mixins/actions/fetch.js (hunk 13 above)
Comment (pascalwengerter) [reply packages/web-app-files/src/mixins/actions/fetch.js:58]: Not sure about the implications, @kulmann ?

File: packages/web-app-files/src/mixins/actions/fetch.js (hunk 12 above)
Comment (pascalwengerter) [reply packages/web-app-files/src/mixins/actions/fetch.js:47]: Not sure about the implications, @kulmann ?

File: tests/acceptance/features/webUIDeleteFilesFolders/deleteFilesFolders.feature (hunk 7 above)
Comment (pascalwengerter) [reply tests/acceptance/features/webUIDeleteFilesFolders/deleteFilesFolders.feature:251]: Done, thanks

File: packages/web-app-files/src/components/ActionMenuItem.vue [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+
+      return {
+        appearance: this.appearance,
+        ...(action.isDisabled && { disabled: action.isDisabled() }),
+        ...(action.variation && { variation: action.variation })
+      }
+    },
+
+    getComponentListeners(action, resources) {
+      if (action.handler === undefined || action.componentType !== 'oc-button') {
```
[the remark this answers is not in the document]
Comment (pascalwengerter) [reply packages/web-app-files/src/components/ActionMenuItem.vue:76]: Aye, thanks!

File: packages/web-app-files/src/mixins/actions/fetch.js (hunk 12 above)
Comment (fschade) [reply packages/web-app-files/src/mixins/actions/fetch.js:45]: to get this merged it say keep the fixme comment here and do it later

**Review by fschade — APPROVED** [review]
🚀 🎸  - LGTM
